### PR TITLE
pipewire: Add new output plugin. Closes: #1048

### PR DESCRIPTION
**Pull request by radioactiveman**
This is my attempt to port the PipeWire output plugin from Qmmp to Audacious.
[Ilya Kotov](https://github.com/trialuser02), author of the related [Qmmp code](https://sourceforge.net/p/qmmp-dev/code/HEAD/tree/branches/qmmp-2.0/src/plugins/Output/pipewire/) which is under GPL, allowed me to use the BSD license for our code base.

I'm using the plugin successfully for some time now, but further tests are of course appreciated. :)
Could you please support with the remaining things listed below? I don't know how to do them.

* Implement `period_wait()`.
* Use the ring buffer from `libaudcore` instead of the raw char array.
* A few variables are still hard-coded, this should be cleaned-up.

Other questions:
- I assume the channel mappings in `set_channel_map()` can be reduced, since Audacious only supports up to [10 channels](https://github.com/audacious-media-player/audacious/blob/e7addd0068ac3cb17fbedc29bd852a53088184ad/src/libaudcore/audio.h.in#L23). Is this correct?
- Is it necessary to include all format mappings in `to_pipewire_format()`? The PulseAudio plugin for example only maps a few formats in `to_pulse_format()`. Why is that?
- Is the priority ok? Should it be lower than the PulseAudio plugin for now?

### CODE REVIEW

**Comment by mcarans** [comment]
Would be great to see this merged.

**Comment by PedroHLC** [comment]
Demo

LGTM, or better, sounding good to me (tested on NixOS).

**Comment by jlindgren90** [comment]
The biggest issue I see here is `period_wait()` isn't implemented. Calling `pw_thread_loop_timed_wait()` from `write_audio()` will potentially block the UI thread and make Audacious as a whole less responsive. Until we find a way to solve that, I would think it would be better to continue using the PulseAudio output plugin, as I believe that should be usable still with PipeWire?

**Comment by jlindgren90** [comment]
@radioactiveman It may enough to just move the `pw_thread_loop_timed_wait()` bit from `write_audio()` to `period_wait()`. Would you be able to make/test that change?  (I am not using PipeWire on any of my own systems currently.)

File: src/pipewire/pipewire.cc [hunk 1]
```diff
@@ -0,0 +1,575 @@
...
+
+    // TODO: Don't hardcode values, use aud_get_int("output_buffer_size")?
+    //
+    // m_stride = AudioParameters::sampleSize(format) * map.count();
+    // m_frames = qBound(64, qCeil(2048.0 * rate / 48000.0), 8192);
+    // m_buffer_size = m_frames * m_stride;
+    // m_buffer = new unsigned char[m_buffer_size];
+    m_stride = 4;
+    m_frames = 128;
+    m_buffer_size = 1024;
```
Comment (jlindgren90) [inline src/pipewire/pipewire.cc:330]: If the commented line above is correct
```
m_buffer_size = m_frames * m_stride;
```

then should this be:
```
m_buffer_size = 512;
```

?

File: src/pipewire/pipewire.cc [hunk 2]
```diff
@@ -0,0 +1,575 @@
...
+    if (!(m_loop = pw_thread_loop_new("pipewire-main-loop", nullptr)))
+    {
+        AUDERR("PipeWireOutput: unable to create main loop\n");
+        return false;
+    }
+
+    if (!(m_context = pw_context_new(pw_thread_loop_get_loop(m_loop), nullptr, 0)))
+    {
+        AUDERR("PipeWireOutput: unable to create context\n");
+        return false;
```
Comment (jlindgren90) [inline src/pipewire/pipewire.cc:283]: Most of the `return false;` paths in this function appear to leak memory/not clean things up properly (e.g. here we leak `m_loop` which was initialized above).

File: src/pipewire/pipewire.cc [hunk 3]
```diff
@@ -0,0 +1,575 @@
...
+{
+    PipeWireOutput * o = static_cast<PipeWireOutput *>(data);
+
+    if (!o->m_buffer_at)
+    {
+        pw_thread_loop_signal(o->m_loop, false);
+        return;
+    }
+
+    struct pw_buffer * b = pw_stream_dequeue_buffer(o->m_stream);
```
Comment (jlindgren90) [inline src/pipewire/pipewire.cc:474]: Can `pw_stream_dequeue_buffer()` return `nullptr`?  The line below will crash if so:
```
struct spa_buffer * buf = b->buffer;
```

**Review by jlindgren90 — CHANGES_REQUESTED** [review]

**Comment by radioactiveman** [comment]
Thanks for the review @jlindgren90 and thanks @PedroHLC for testing. :+1:
I will try to address the remarks in the next days when I have some time.

File: src/pipewire/pipewire.cc (hunk 2 above)
Comment (radioactiveman) [reply src/pipewire/pipewire.cc:306]: Fixed by calling `close_audio()` in these cases, similar to the PulseAudio plugin.

File: src/pipewire/pipewire.cc [hunk 4]
```diff
@@ -0,0 +1,597 @@
...
+        values[0] = v.left / 100.0f;
+        values[1] = v.right / 100.0f;
+    }
+    else
+    {
+        for (unsigned int i = 0; i < m_channels; i++)
+            values[i] = aud::max(v.left, v.right) / 100.0f;
+    }
+
+    pw_thread_loop_lock(m_loop);
```
Comment (radioactiveman) [inline src/pipewire/pipewire.cc:153]: Should the `lock` be done at the beginning of the method? Or does it not matter here? Same question for `init_stream()`.

File: src/pipewire/pipewire.cc [hunk 5]
```diff
@@ -0,0 +1,597 @@
...
+    pw_thread_loop_signal(o->m_loop, false);
+}
+
+void PipeWireOutput::on_drained(void * data)
+{
+    PipeWireOutput * o = static_cast<PipeWireOutput *>(data);
+    pw_thread_loop_signal(o->m_loop, false);
+}
+
+enum spa_audio_format PipeWireOutput::to_pipewire_format(int format)
```
Comment (radioactiveman) [inline src/pipewire/pipewire.cc:535]: Do we need to handle all formats here? The PulseAudio plugin checks less in [to_pulse_format()](https://github.com/audacious-media-player/audacious-plugins/blob/45a0df0ecb479a64d85f982b8cc087b7547e5562/src/pulse/pulse_audio.cc#L369).

**Comment by radioactiveman** [comment]
> @radioactiveman It may enough to just move the `pw_thread_loop_timed_wait()` bit from `write_audio()` to `period_wait()`. Would you be able to make/test that change?

Yes, this works. :+1: And I have noticed that the volume slider in in fact more responsive now, especially with the GTK interface.

**Comment by radioactiveman** [comment]
@jlindgren90: I have rebased this to `master` and addressed your review comments. CI is unfortunately not possible since the PipeWire package is too old on Ubuntu 20.04.

Could you please answer my two questions in the review? Thanks and looking forward to the merge. :)

File: src/pipewire/pipewire.cc (hunk 4 above)
Comment (jlindgren90) [reply src/pipewire/pipewire.cc:153]: I think they are fine where they are :+1:

File: src/pipewire/pipewire.cc (hunk 5 above)
Comment (jlindgren90) [reply src/pipewire/pipewire.cc:535]: I don't think we ever send 8-bit or unsigned formats to the output plugins. Supporting them doesn't do any harm though.

**Review by jlindgren90 — APPROVED** [review]

**Comment by jlindgren90** [comment]
@radioactiveman Looks good to me. Thanks for all your work on this!

**Comment by birdie-github** [comment]
PipeWire uses/offers _the exact same API_ as PulseAudio, in other words PA plugin is _100% compatible and sufficient_ for PipeWire.

There's _no_ need to have two plugins. You may rename the old one from "PulseAudio plugin" to "PulseAudio/PipeWire plugin" and that's it.

**Comment by kaniini** [comment]
One plugin uses libpulse, and the other uses libpipewire directly.  There's no need to support either sound system, but we do anyway.

**Pull request merged**